[KEP - 3085] Add e2e test for PodReadyToStartContainerCondition when config map is created

charles-chenzz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind flake
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
add e2e test for podreadytostartcontainercondition, as this is part of KEP 3085
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #121317

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

charles-chenzz committed:
update testcase on podreadytostartcontainercondition

charles-chenzz commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -137,20 +138,27 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+			},
+		}
+		e2epod.NewPodClient(f).Apply(ctx, applyConfig, metav1.ApplyOptions{})
+
```
the original test was flaked due to I use update to create volume which is not allowed. using apply should be able to fix the flake

kannon92 replied:
Can you do this with `Kind` or a k8 cluster?  ie update a volume to an existing pod.  Usually pod updates are pretty locked down.

kannon92 commented:
@charles-chenzz I think we should revert the previous PR.  

I don't think this test actually does what we want.  I am also not exactly sure if we can do this after all.  

The test case I was thinking of:

1) Create a Pod with volumes pointed to a secret/config-map
2) Verify that PodReadyToStartContainers is false due to not existing secret/config-map
3) Create secret/config-map 
4) Wait for pod to have PodReadyToStartContainers true.

I don't think you are able to add volumes to a pod after it has been running.  I think you want to look into how to create a secret/config-map in the node_e2e tests.

charles-chenzz commented:
> I don't think you are able to add volumes to a pod after it has been running. I think you want to look into how to create a secret/config-map in the node_e2e tests.

can you point out in which files?

I think we should wait the CI hints and gain more infos then decide if we need to revert the previous PR. wdyt? @kannon92

kannon92 commented:
> > I don't think you are able to add volumes to a pod after it has been running. I think you want to look into how to create a secret/config-map in the node_e2e tests.
> 
> can you point out in which files?
> 
> I think we should wait the CI hints and gain more infos then decide if we need to revert the previous PR. wdyt? @kannon92

I can't actually find a test in node_e2e that creates a secret or config-map.  So I am thinking that this test may be tricky than we thought.

I opened up a revert just in case

kannon92 commented:
/retitle WIP: Investigate flaky test for PodReadyToStartContainerCondition

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

aojea commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -206,6 +242,10 @@ func getTransitionTimeForPodConditionWithStatus(pod *v1.Pod, condType v1.PodCond
 
 func webserverPodSpec(podName, containerName, initContainerName string, addInitContainer bool) *v1.Pod {
 	p := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
```
this is not needed

charles-chenzz replied:
good catch. will update soon. I was setting this field while investigating how to use ownerReference and localObjectReferece to combine the configmap we missing here.

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,30 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+		// Verify PodReadyToStartContainers is set (since sandboxcreation is unblocked)
+		if checkPodReadyToStart {
+			_, err = getTransitionTimeForPodConditionWithStatus(p, v1.PodReadyToStartContainers, true)
+			framework.ExpectNoError(err)
```
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121321/pull-kubernetes-node-kubelet-serial-containerd/1715962446105022464

the configmap is created but the status is still false, been debug for a while and couldn't find out. any ideas? @kannon92

charles-chenzz replied:
update with check in a eventually call. let's give it a try

charles-chenzz replied:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121321/pull-kubernetes-node-kubelet-serial-containerd/1716241761388466176

still couldn't fetch the status. I'll try recreating the pod after cm is up

kannon92 replied:
Yea I think what we would need to do is verify that the pod is recreated/restarted but I am not entirely sure how the pod lifecycle works.  

Recreating should work.

charles-chenzz replied:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121321/pull-kubernetes-node-kubelet-serial-containerd/1716422537929297920

still not work I think. plus there's some unrelated error from the graceful-shutdown. however, I would try another method by using createSync(don't know if it works). but still, let's have a try

kannon92 replied:
We followed offline for this and fixed this. If I remember right, the fix was making sure that configMap was in the same namespace.

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

kannon92 commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,55 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+			},
+		}
+
+		p2 = e2epod.NewPodClient(f).Create(ctx, p2)
```
I think you want to create the config map before you create this pod.  

1) create configmap
2) create pod that references config-map
3) verify true condition.

If that doesn't work, I'd suggest trying to test this flow manually and see if there is something wrong.

charles-chenzz replied:
move the configmap create before pod. waiting for CI.

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

kannon92 commented:
I tested this flow (create config map, create pod that references).  It seems to work fine so I think there is something up with the e2e tests. I can help look at this on Monday.

kannon92 commented:
```diff
diff --git a/test/e2e_node/pod_conditions_test.go b/test/e2e_node/pod_conditions_test.go
index 3a6cdc1b19b..9767f9fd412 100644
--- a/test/e2e_node/pod_conditions_test.go
+++ b/test/e2e_node/pod_conditions_test.go
@@ -149,11 +149,15 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
                        },
                }

-               _, err = f.ClientSet.CoreV1().ConfigMaps("default").Create(ctx, &configmap, metav1.CreateOptions{})
+               _, err = f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, &configmap, metav1.CreateOptions{})
                framework.ExpectNoError(err)

-               _, err = f.ClientSet.CoreV1().ConfigMaps("default").Get(ctx, "cm-does-not-exist", metav1.GetOptions{})
+               _, err = f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Get(ctx, "cm-does-not-exist", metav1.GetOptions{})
                framework.ExpectNoError(err)
+               defer func() {
+                       err := f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Delete(ctx, "cm-does-not-exist", metav1.DeleteOptions{})
+                       framework.ExpectNoError(err, "not able to delete configmap")
+               }()

                p2 := webserverPodSpec("pod2-"+string(uuid.NewUUID()), "web2", "init2", hasInitContainers)
                p2.Spec.Volumes = []v1.Volume{
@@ -174,15 +178,15 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
                }

                p2 = e2epod.NewPodClient(f).Create(ctx, p2)
-               framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, f.ClientSet, p.Name, f.Namespace.Name, framework.PodStartTimeout))
+               framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, f.ClientSet, p2.Name, p2.Namespace, framework.PodStartTimeout))

                p2, err = e2epod.NewPodClient(f).Get(ctx, p2.Name, metav1.GetOptions{})
                framework.ExpectNoError(err)

-               _, err = getTransitionTimeForPodConditionWithStatus(p, v1.PodScheduled, true)
+               _, err = getTransitionTimeForPodConditionWithStatus(p2, v1.PodScheduled, true)
                framework.ExpectNoError(err)

-               _, err = getTransitionTimeForPodConditionWithStatus(p, v1.PodInitialized, true)
+               _, err = getTransitionTimeForPodConditionWithStatus(p2, v1.PodInitialized, true)
                framework.ExpectNoError(err)

                // Verify PodReadyToStartContainers is set (since sandboxcreation is unblocked)
```

This patch should fix this test.  I ran it locally and I was able to get these jobs to pass.  

Summarizing these changes:

1) Config maps should be deleted after each and every test.
2) Config maps are not created in the default namespace but a test namespace.
3) The last checks should use the new pod that you created.

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
the fails seems unrelated to the changes we made here.

charles-chenzz commented:
/test pull-kubernetes-unit

kannon92 commented:
/retitle Add e2e test for PodReadyToStartContainerCondition when config map is created

kannon92 commented:
@charles-chenzz I think this is good to go.  Didn't see these failures anymore.

rphillips commented:
/triage accepted
/priority important-soon
/lgtm

charles-chenzz commented:
> @charles-chenzz can you update the description of this PR? We are including this as part of the beta promotion for PodReadyToStartContainers. I added it to the KEP issue already but it would be nice for the description to reflect that.

added on the desc.

charles-chenzz commented:
/remove-kind flake

kannon92 commented:
/kind feature

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
the fails seems unrelated to the changes

kannon92 commented:
/retitle [KEP -3085] Add e2e test for PodReadyToStartContainerCondition when config map is created

kannon92 commented:
/retitle [KEP - 3085] Add e2e test for PodReadyToStartContainerCondition when config map is created

dims commented:
/assign @tzneal @mrunalp @SergeyKanzhelev

SergeyKanzhelev commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,67 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+
+		configmap := v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cm-does-not-exist",
```
why is it called `cm-does-not-exist` if it exists? Either I don't understand the name or the what that test validates. Can you please clarify?

charles-chenzz replied:
the naming of this cm is wrong..
the test try to check pod condition for a pod when volumes is created( it's blocked due to cm doesn't exist). it tends to verify ``PodReadyToStartContainer`` condition have boolean change after the volume(configmap here)  create

charles-chenzz replied:
renamed. please take a look when you get a chance. @SergeyKanzhelev

kannon92 replied:
@charles-chenzz please take a look at this again. I think you didn't rename all the instances of this.

The test failures seem related now.

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

kannon92 commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,62 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+		_, err = f.ClientSet.CoreV1().ConfigMaps("default").Create(ctx, &configmap, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		_, err = f.ClientSet.CoreV1().ConfigMaps("default").Get(ctx, "cm-does-not-exist", metav1.GetOptions{})
```
This probably should be cm-that-unblock-pod-condition rather than cm-does-not-exist.

kannon92 commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,62 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+				Name: "cm-2",
+				VolumeSource: v1.VolumeSource{
+					ConfigMap: &v1.ConfigMapVolumeSource{
+						LocalObjectReference: v1.LocalObjectReference{Name: "cm-does-not-exist"},
```
Same here. This config map doesn't exist anymore.

charles-chenzz replied:
my mistake. will update later today

kannon92 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

kannon92 commented:
The PodReadyToStart test cases are successful.

Thank you @charles-chenzz 

/lgtm
/assign @SergeyKanzhelev

kannon92 commented:
For approvers, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121321/pull-kubernetes-node-kubelet-serial-containerd/1745075720918405120 has some known failures. I was making sure that this feature does not flake. This looks good to me.

pacoxu commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,67 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+		_, err = f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, &configmap, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		_, err = f.ClientSet.CoreV1().ConfigMaps(f.Namespace.Name).Get(ctx, "cm-that-unblock-pod-condition", metav1.GetOptions{})
+		framework.ExpectNoError(err)
```
If the create succeed, we don't have to get it.

```suggestion
```

kannon92 replied:
cc @charles-chenzz

charles-chenzz replied:
@kannon92 paco was right, I get it here to check it explicitly during research and now we can confirm that it's created, we can remove it

pacoxu commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,67 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
 		_, err = getTransitionTimeForPodConditionWithStatus(p, v1.PodReady, false)
 		framework.ExpectNoError(err)
+
+		ginkgo.By("checking pod condition for a pod when volumes is created")
```
```suggestion
		ginkgo.By("checking pod condition for a pod with a created configmap volume")
```
This seems to be a success conditions test. This function is called `runPodFailingConditionsTest`. 

Move this part to `runPodReadyConditionsTest()` below.

charles-chenzz replied:
this test case need to success after the upper one was failed. can we move it to ``runPodReadyConditionsTest`` ?

pacoxu replied:
Or we may just rename the current `runPodFailingConditionsTest` 😄. (or just a comment for the function)

charles-chenzz replied:
can we add comment for L137 to specifically express the purpose?

pacoxu reviewed (changes requested):
Or just add the configmap volume for the test case in `runPodReadyConditionsTest` and also the check there.

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
error was due to my mistake of removing wrong code 😓 fixing later today

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

charles-chenzz commented:
error unrelated now

kannon92 commented:
/lgtm
/assign @pacoxu @SergeyKanzhelev

pacoxu commented on test/e2e_node/pod_conditions_test.go:
```diff
@@ -134,6 +134,66 @@ func runPodFailingConditionsTest(f *framework.Framework, hasInitContainers, chec
+
+		// this testcase is creating the missing volume that unblock the pod above,
+		// and check PodReadyToStartContainer is setting correctly.
+		ginkgo.By("checking pod condition for a pod when volumes is created")
```
```suggestion
		ginkgo.By("checking pod condition for a pod when volume source is created")
```

kannon92 replied:
@charles-chenzz while we are waiting for approval, could you apply this change?

charles-chenzz replied:
applied

pacoxu reviewed (approved):
/lgtm
one nit 🤣

/assign @mrunalp

kannon92 commented:
cc @SergeyKanzhelev @mrunalp

charles-chenzz commented:
/test pull-kubernetes-node-kubelet-serial-containerd

kannon92 commented:
/lgtm

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

kannon92 commented:
Nice work on leading this @charles-chenzz!
